src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Acquisition/OptionalWorkloadProjectLoadVeto.cs

```diff
@@ -0,0 +1,107 @@
+        public OptionalWorkloadProjectLoadVeto(IProjectAccessor projectAccessor,
+            UnconfiguredProject unconfiguredProject,
+            IVsService<SVsBrokeredServiceContainer, IBrokeredServiceContainer> serviceBrokerContainer)
+        {
+            _projectAccessor = projectAccessor;
+            _unconfiguredProject = unconfiguredProject;
+            _serviceBrokerContainer = serviceBrokerContainer;
+        }
+
+        /// <summary>
+        /// Checks whether the project needs optional workloads (that are not installed)
+        /// using the evaluation model. The purpose of this check is to reject loads
+        /// of such projects based on the evaluation of the initial project configuration.
+        /// </summary>
+        async Task<bool> IVetoProjectLoad.AllowProjectLoadAsync(bool isNewProject, CancellationToken cancellationToken)
+        {
+            ConfiguredProject? configuredProject = await _unconfiguredProject.GetSuggestedConfiguredProjectAsync();
+            if (configuredProject == null)
+            {
+                return true;
+            }
+
+            var serviceBrokerContainer = await _serviceBrokerContainer.GetValueAsync(cancellationToken);
+            IServiceBroker? serviceBroker = serviceBrokerContainer?.GetFullAccessServiceBroker();
+            if (serviceBroker == null)
+            {
+                return true;
+            }
+
+            var missingWorkloadRegistrationService = await serviceBroker.GetProxyAsync<IMissingWorkloadRegistrationService>(
+                serviceDescriptor: VisualStudioServices.VS2022.MissingWorkloadRegistrationService,
+                cancellationToken: cancellationToken);
+
+            using (missingWorkloadRegistrationService as IDisposable)
+            {
+                if (missingWorkloadRegistrationService == null)
+                {
+                    return true;
+                }
+
+                var projectGuid = await _unconfiguredProject.GetProjectGuidAsync();
+
+                try
+                {
+                    return await _projectAccessor.OpenProjectForReadAsync(configuredProject, evaluatedProject =>
+                    {
+                        var missingWorkloadComponentIds = GetVSComponentIdsForMissingWorkloads(evaluatedProject);
+
+                        if (missingWorkloadComponentIds.Any())
+                        {
+                            missingWorkloadRegistrationService.RegisterMissingWorkloadComponentIds(projectGuid, missingWorkloadComponentIds);
+                            return false;
+                        }
+
+                        return true;
+                    },
+                    cancellationToken);
+                }
+                finally
+                {
+                    (missingWorkloadRegistrationService as IDisposable)?.Dispose();
+                }
+            }
+        }
+
+        private IEnumerable<string> GetVSComponentIdsForMissingWorkloads(Project project)
```

Comment:
Consider returning `IReadOnlyList<>` here to convey that it's safe for consuming code to enumerate multiple times (as it currently does).